src/error.rs

```diff
@@ -31,18 +39,56 @@ impl std::fmt::Display for Error {
 }
 
 /// The types of errors that can occur in the `bson` crate.
-#[derive(Debug, Error)]
+#[derive(Clone, Debug, Error)]
 #[non_exhaustive]
 pub enum ErrorKind {
+    /// An error related to the [`Binary`](crate::Binary) type occurred.
```

Comment:
This is exposing how arbitrary it is which of our errors carry a string and which carry a subkind enum; that's got a bit of a code smell to it but I can't decide if it's _actually_ a problem worth addressing.  Do you have thoughts here?

Reply:
Yeah, I don't _love_ this design; I agree that it's arbitrary and inconsistent. My concern, however, is that users who do have some kind of programmatic use for the values in these errors will be less inclined to migrate to the new version, so my instinct here is to err on the side of portability for lower-priority changes. Certainly open to other designs here if you had something else in mind though!

Reply:
Sketch of a proposal:
* Hoist `message` to an optional field of `Error`, with corresponding code in the `Display` impl to append it if it's present.
* All variants of `ErrorKind` are required to be `#[non_exhaustive]` struct variants, even if they're empty (which a lot now will be with `message` stripped out)
* Any sub-`Kind` types follow the same pattern (sub-`Kinds` also don't need to have `message` since they're ultimately attached to an `Error`!)

This makes the error space a lot more uniform, gives us a lot of flexibility for future error evolution, and doesn't introduce substantial migration burden.